File: py/src/aoc2020/machine.py
from dataclasses import dataclass
from enum import Enum


class Operation(Enum):
    ACC = 1
    JMP = 2
    NOP = 3


@dataclass
class Instruction:
    op: Operation
    value: int

    def __init__(self, line):
        if line.startswith('acc '):
            self.op = Operation.ACC
        elif line.startswith('jmp '):
            self.op = Operation.JMP
        elif line.startswith('nop '):
            self.op = Operation.NOP
        else:
            raise ValueError(f'Invalid instruction: {line!r}')
        self.value = int(line[4:].removeprefix('+').rstrip())

    def __iter__(self):
        return iter((self.op, self.value))


class Machine:
    def __init__(self, instructions):
        self._instructions = instructions
        self._acc = 0
        self._ip = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self._ip not in range(len(self._instructions)):
            raise StopIteration()
        op, value = self._instructions[self._ip]
        if op == Operation.NOP:
            self._ip += 1
        elif op == Operation.JMP:
            self._ip += value
        elif op == Operation.ACC:
            self._acc += value
            self._ip += 1
        else:
            raise ValueError(f'Invalid operation: {operation!r}')
        return self._acc, self._ip

File: rs/src/machine.rs
use std::convert::TryInto;
use std::error::Error;
use std::fmt::{self, Display, Formatter};
use std::ops::AddAssign;
use std::str::FromStr;

#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub enum ParseInstructionErr<E> {
    Format,
    Operation(String),
    Value(E),
}

impl<E> Display for ParseInstructionErr<E>
where
    E: Display,
{
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        match self {
            Self::Format => write!(f, "ParseInstructionErr::Format"),
            Self::Operation(op) => write!(f, "ParseInstructionErr::Operation({})", op),
            Self::Value(err) => write!(f, "ParseInstructionErr::Value({})", err),
        }
    }
}

impl<E> Error for ParseInstructionErr<E> where E: Error {}

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub enum Instruction<T> {
    Acc(T),
    Jmp(T),
    Nop(T),
}

impl<T> FromStr for Instruction<T>
where
    T: FromStr,
{
    type Err = ParseInstructionErr<<T as FromStr>::Err>;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if let Some(value) = s.strip_prefix("acc ") {
            Ok(Self::Acc(value.parse().map_err(Self::Err::Value)?))
        } else if let Some(value) = s.strip_prefix("jmp ") {
            Ok(Self::Jmp(value.parse().map_err(Self::Err::Value)?))
        } else if let Some(value) = s.strip_prefix("nop ") {
            Ok(Self::Nop(value.parse().map_err(Self::Err::Value)?))
        } else {
            Err(s
                .split(' ')
                .next()
                .map_or(Self::Err::Format, |op| Self::Err::Operation(op.to_string())))
        }
    }
}

#[derive(Debug)]
pub struct Machine<'a, T> {
    instructions: &'a [Instruction<T>],
    acc: T,
    ip: Option<usize>,
}

impl<'a, T> Machine<'a, T>
where
    T: Default,
{
    pub fn new(instructions: &'a [Instruction<T>]) -> Machine<'a, T> {
        Machine {
            instructions,
            acc: T::default(),
            ip: if instructions.is_empty() {
                None
            } else {
                Some(0)
            },
        }
    }
}

impl<'a, T> Iterator for Machine<'a, T>
where
    T: AddAssign,
    T: Copy,
    T: Into<isize>,
{
    type Item = (T, Option<usize>);
    fn next(&mut self) -> Option<Self::Item> {
        let ip = self.ip?;
        let mut jmp = 1;
        match &self.instructions[ip] {
            Instruction::Acc(x) => self.acc += *x,
            Instruction::Jmp(x) => jmp = (*x).into(),
            Instruction::Nop(_) => {}
        }
        self.ip = ip
            .try_into()
            .ok()
            .and_then(|ip: isize| ip.checked_add(jmp)?.try_into().ok())
            .filter(|ip| *ip < self.instructions.len());
        Some((self.acc, self.ip))
    }
}
